In [1]:
import pandas as pd
import sklearn as sn


In [5]:
df= pd.read_csv("C:\\Users\\Maddy\\Downloads\\store.csv")

In [6]:
df

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
1     5575-GNVDE    Male              0      No         No      34   
2     3668-QPYBK    Male              0      No         No       2   
3     7795-CFOCW    Male              0      No         No      45   
4     9237-HQITU  Female              0      No         No       2   
...          ...     ...            ...     ...        ...     ...   
7038  6840-RESVB    Male              0     Yes        Yes      24   
7039  2234-XADUH  Female              0     Yes        Yes      72   
7040  4801-JZAZL  Female              0     Yes        Yes      11   
7041  8361-LTMKD    Male              1     Yes         No       4   
7042  3186-AJIEK    Male              0      No         No      66   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
0              No  No phone service             DSL             No  ...   
1             Yes                No             DSL            Yes  ...   
2             Yes                No             DSL            Yes  ...   
3              No  No phone service             DSL            Yes  ...   
4             Yes                No     Fiber optic             No  ...   
...           ...               ...             ...            ...  ...   
7038          Yes               Yes             DSL            Yes  ...   
7039          Yes               Yes     Fiber optic             No  ...   
7040           No  No phone service             DSL            Yes  ...   
7041          Yes               Yes     Fiber optic             No  ...   
7042          Yes                No     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No          No          No              No  Month-to-month   
1                 Yes          No          No              No        One year   
2                  No          No          No              No  Month-to-month   
3                 Yes         Yes          No              No        One year   
4                  No          No          No              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7038              Yes         Yes         Yes             Yes        One year   
7039              Yes          No         Yes             Yes        One year   
7040               No          No          No              No  Month-to-month   
7041               No          No          No              No  Month-to-month   
7042              Yes         Yes         Yes             Yes        Two year   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
0                 Yes           Electronic check          29.85         29.85   
1                  No               Mailed check          56.95        1889.5   
2                 Yes               Mailed check          53.85        108.15   
3                  No  Bank transfer (automatic)          42.30       1840.75   
4                 Yes           Electronic check          70.70        151.65   
...               ...                        ...            ...           ...   
7038              Yes               Mailed check          84.80        1990.5   
7039              Yes    Credit card (automatic)         103.20        7362.9   
7040              Yes           Electronic check          29.60        346.45   
7041              Yes               Mailed check          74.40         306.6   
7042              Yes  Bank transfer (automatic)         105.65        6844.5   

     Churn  
0       No  
1       No  
2      Yes  
3       No  
4      Yes  
...    ...  
7038    No  
7039    No  
7040    No  
7041   Yes  
7042    No  

[7043 rows x 21 columns]

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


# What is the average tenure of customers who churn?

In [12]:
df.groupby("Churn").mean('tenure')

SeniorCitizen     tenure  MonthlyCharges
Churn                                          
No          0.128721  37.569965       61.265124
Yes         0.254682  17.979133       74.441332

## hence the average tenure of customers who churn is 17.979133

# What is the correlation between gender and churn?

In [88]:
def binary_map(feature):
    return feature.map({'Yes':1, 'No':0})

In [98]:
q2=df[['gender','Churn']]

In [99]:
q2['Churn'] = q2[['Churn']].apply(binary_map)

C:\Users\Maddy\AppData\Local\Temp\ipykernel_10220\2138188848.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  q2['Churn'] = q2[['Churn']].apply(binary_map)


In [100]:
q2['gender'] = q2['gender'].map({'Male':1, 'Female':0})

C:\Users\Maddy\AppData\Local\Temp\ipykernel_10220\267084904.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  q2['gender'] = q2['gender'].map({'Male':1, 'Female':0})


In [101]:
q2

gender  Churn
0          0      0
1          1      0
2          1      1
3          1      0
4          0      1
...      ...    ...
7038       1      0
7039       0      0
7040       0      0
7041       1      1
7042       1      0

[7043 rows x 2 columns]

In [ ]:
q2.c

In [42]:
q2.gender.replace("Female",1,inplace=True)
q2.gender.replace("Male",1,inplace=True)

C:\Users\Maddy\AppData\Local\Temp\ipykernel_10220\2365922045.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  q2.gender.replace("Female",1,inplace=True)
C:\Users\Maddy\AppData\Local\Temp\ipykernel_10220\2365922045.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  q2.gender.replace("Male",1,inplace=True)


In [20]:
df.Churn.unique()

array(['No', 'Yes'], dtype=object)

# What is the difference in the average monthly charges between customers who churn and those who do not?


In [50]:
q3=df.groupby("Churn").mean("MonthlyCharges")

In [52]:
q3.MonthlyCharges.diff()

Churn
No           NaN
Yes    13.176209
Name: MonthlyCharges, dtype: float64

# Average monthly charges between customers who churn and those who do not is 13.176209

# What are the most popular internet services usage type among customers who churn?


In [74]:
q4=df[['Churn',"InternetService"]]

In [84]:
q4=q4.loc[q4['Churn']=='Yes']

In [85]:
q4.groupby("InternetService").count()

Churn
InternetService       
DSL                459
Fiber optic       1297
No                 113


# Fiber optic is preferred among the churn customer

# What is the trend in churn over time?

## There is no Date and time column to find the trend over time

# What are the factors that are most important in determining whether a customer will churn?

The three leading factors that impact customer churn rate:

1. Average subscription length Subscription length is the amount of time an average customer spends paying for a company’s goods or services. Typical subscription length varies based on the type of service offered. ...
2. Customer acquisition cost Another crucial factor in churn rate is customer acquisition cost, or the amount of money spent to gain one new customer. ...
3. Customer lifetime value (CLV)